torch/nn/parallel/comm.py

```diff
@@ -0,0 +1,166 @@
+import torch
+from torch.cuda import nccl
+from torch._utils import _take_tensors, _flatten_dense_tensors, \
+    _unflatten_dense_tensors, _reorder_tensors_as, _get_current_device_index
+
+
+def broadcast(tensor, devices):
+    """Broadcasts a tensor to a number of GPUs.
+
+    Arguments:
+        tensor (Tensor): tensor to broadcast.
+        devices (Iterable): an iterable of devices among which to broadcast.
+          Note that it should be like (src, dst1, dst2, ...), the first element
+          of which is the source device to broadcast from.
+
+    Returns:
+        A tuple containing copies of the ``tensor``, placed on devices
+        corresponding to indices from ``devices``.
+    """
+    return torch._C._broadcast(tensor, devices)
+
+
+def broadcast_coalesced(tensors, devices, buffer_size=10485760):
+    """Broadcasts a sequence tensors to the specified GPUs.
+    Small tensors are first coalesced into a buffer to reduce the number
+    of synchronizations.
+
+    Arguments:
+        tensors (sequence): tensors to broadcast.
+        devices (Iterable): an iterable of devices among which to broadcast.
+          Note that it should be like (src, dst1, dst2, ...), the first element
+          of which is the source device to broadcast from.
+        buffer_size (int): maximum size of the buffer used for coalescing
+
+    Returns:
+        A tuple containing copies of the ``tensor``, placed on devices
+        corresponding to indices from ``devices``.
+    """
+    return torch._C._broadcast_coalesced(tensors, devices, buffer_size)
+
+
+def reduce_add(inputs, destination=None):
+    """Sums tensors from multiple GPUs.
+
+    All inputs should have matching shapes.
+
+    Arguments:
+        inputs (Iterable[Tensor]): an iterable of tensors to add.
+        destination (int, optional): a device on which the output will be
+            placed (default: current device).
+
+    Returns:
+        A tensor containing an elementwise sum of all inputs, placed on the
+        ``destination`` device.
+    """
+    # TODO: try to find an input on another gpu, copy it,
+    device_type = inputs[0].device.type
+    # and accumulate into the copy
+    if destination is None:
+        destination = _get_current_device_index()
```

Comment:
I saw three changes in this file:

1. this line
2. `inp.is_cuda` -> `inp.device.type != "cpu"`
3. `inp.cuda(result.get_device())` -> `inp.to(result.device)`

If these are the only changes, it seems we can simply let `cuda/comm.py` call into this one, and then we will also get tests for free?

Reply:
Sounds good.
The cuda/comm.py can be a shim-layer for backward compatibility.